backend/danswer/direct_qa/request_model.py

```diff
@@ -0,0 +1,201 @@
+        api_key: str | None,
+        max_output_tokens: int,
+        stream: bool,
+        timeout: int | None,
+    ) -> Response:
+        """Given a filled out prompt, how to send it to the model API with the
+        correct request format with the correct parameters"""
+        raise NotImplementedError
+
+    @staticmethod
+    @abc.abstractmethod
+    def extract_model_output_from_response(
+        response: Response,
+    ) -> str:
+        """Extract the full model output text from a response.
+        This is for nonstreaming endpoints"""
+        raise NotImplementedError
+
+    @staticmethod
+    @abc.abstractmethod
+    def generate_model_tokens_from_response(
+        response: Response,
+    ) -> Generator[str, None, None]:
+        """Generate tokens from a streaming response
+        This is for streaming endpoints"""
+        raise NotImplementedError
+
+
+class HuggingFaceRequestModel(HostSpecificRequestModel):
+    @staticmethod
+    def send_model_request(
+        filled_prompt: str,
+        endpoint: str,
+        api_key: str | None,
+        max_output_tokens: int,
+        stream: bool,  # Not supported by Inference Endpoints (as of Aug 2023)
+        timeout: int | None,
+    ) -> Response:
+        headers = {
+            "Authorization": f"Bearer {api_key}",
+            "Content-Type": "application/json",
+        }
+
+        data = {
+            "inputs": filled_prompt,
+            "parameters": {
+                # HuggingFace requires this to be strictly positive from 0.0-100.0 noninclusive
+                "temperature": 0.01,
+                # Skip the long tail
+                "top_p": 0.9,
+                "max_new_tokens": max_output_tokens,
+            },
+        }
+        try:
+            return requests.post(endpoint, headers=headers, json=data, timeout=timeout)
+        except TimeoutError as error:
+            raise TimeoutError(f"Model inference to {endpoint} timed out") from error
+
+    @staticmethod
+    def _hf_extract_model_output(
+        response: Response,
+    ) -> str:
+        if response.status_code != 200:
+            response.raise_for_status()
+
+        return json.loads(response.content)[0].get("generated_text", "")
+
+    @staticmethod
+    def extract_model_output_from_response(
+        response: Response,
+    ) -> str:
+        return HuggingFaceRequestModel._hf_extract_model_output(response)
+
+    @staticmethod
+    def generate_model_tokens_from_response(
+        response: Response,
+    ) -> Generator[str, None, None]:
+        """HF endpoints do not do streaming currently so this function will
+        simulate streaming for the meantime but will need to be replaced in
+        the future once streaming is enabled."""
+        model_out = HuggingFaceRequestModel._hf_extract_model_output(response)
+        yield from simulate_streaming_response(model_out)
+
+
+def get_host_specific_model_class(model_host_type: str) -> HostSpecificRequestModel:
```

Comment:
This is a fantastic opportunity for pluggable, request-based model classes. Having this dynamically load classes when passed a python module path would make it awesome for folks to make plugins that act as pseudo-clients for any kind of inference API. Like VLLM, or Llama.cpp

No changes needed here today, but I might propose a modification in the future for dynamically loaded HostSpecificRequestModel classes.